localstack/services/s3/utils.py

```diff
@@ -196,3 +204,55 @@ def _create_invalid_argument_exc(
 
 def capitalize_header_name_from_snake_case(header_name: str) -> str:
     return "-".join([part.capitalize() for part in header_name.split("-")])
+
+
+def validate_kms_key_id(kms_key: str, bucket: FakeBucket):
+    """
+    Validate that the KMS key used to encrypt the object is valid
+    :param kms_key: the KMS key id or ARN
+    :param bucket: the targeted bucket
+    :raise
+    :return:
+    """
+    try:
+        parsed_arn = parse_arn(kms_key)
+        key_region = parsed_arn["region"]
+        key_id = parsed_arn["resource"].split("/")[1]
+        if key_region != bucket.region_name:
+            raise CommonServiceException(
+                code="KMS.NotFoundException", message=f"Invalid arn {key_region}"
+            )
+
+    except InvalidArnException:
+        # if it fails, the passed ID is a UUID with no region data
+        key_id = kms_key
+        # recreate the ARN manually with the bucket region and bucket owner
+        # if the KMS key is cross-account, user should provide an ARN and not a KeyId
+        kms_key = f"arn:aws:kms:{bucket.region_name}:{bucket.account_id}:key/{key_id}"
```

Comment:
Nit: maybe we can use `localstack.utils.kms_key_arn(key_id, bucket.account_id, bucket.region_name)` here to save some duplication.